### DISCUSSION TITLE
for문과 stream 비교

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#56] [Java]
이상이 알려준거 대답못하면 이상이 속상해합니다.

**Comment by kunsanglee** [comment]
for문은 외부반복을 사용하여 어떻게 할 것인지 하나씩 나열하는 명령형 프로그래밍 방식입니다. 또 추가적인 추상화 레이어가 없기 때문에, 작은 데이터셋에 대해 최적의 성능을 낼 수 있습니다. for문은 추가적인 함수 호출이나 객체 생성 없이 직접적인 데이터 접근을 제공하기 때문에, 메모리 사용량이 상대적으로 낮습니다.

stream은 내부 반복을 사용하여 무엇을 할 것인지에 집중하는 선언형 프로그래밍 방식입니다. stream은 내부적으로 더 많은 추상화와 함수 호출을 포함하기 때문에, 작은 데이터셋에서는 for문에 비해 성능이 떨어질 수 있습니다. 그러나 큰 데이터셋에 대해서는 병렬 스트림을 사용하여 성능 이점을 얻을 수 있습니다. 스트림은 중간 연산결과를 저장하기 위한 추가적인 메모리를 필요로 합니다. 또한 stream은 불변성을 유지하기 위해 새로운 컬렉션을 생성할 수 있으며, 메모리 사용량을 증가시킬 수 있습니다.

**Comment by seunghye218** [comment]
스트림은 보다 기능적이고 선언적인 접근 방식을 제공하여 간결하고 가독성을 높입니다.  
반면 루프는 익숙하고 명시적인 제어 구조를 제공합니다. 정확한 실행 순서나 변경이 필요한 경우에 적합합니다.

성능상으로 일반적으로 for문이 더 빠르지만 병렬 스트림의 경우 다를 수 있습니다.

가독성은 스트림의 선언적 특성 덕분에 읽기 쉬운 코드가 되는 경우가 많습니다. 수행 방법보다 수행해야 할 작업에 더 집중하기 때문입니다.  
반복문은 대부분의 개발자 접근 방식을 잘 이해하고 있습니다. 그러나 코드가 더 장황하여 복잡한 작업의 경우 읽기 어렵습니다.

병렬성과 동시성의 경우,  
스트림은 parallelStream() 메서드 하나로 병렬화 가능합니다.  
for 루프는 세세한 스레드 관리 및 조정이 필요한 복잡한 상황에 적합합니다.

데이터를 수정하는 경우, 스트림에서는 컬렉션 내의 요소를 직접 수정하지 않고 새로운 스트림이나 컬렉션을 생성합니다.
for 루프는 원본을 직접 수정합니다. 이 방법은 기존 데이터를 수정해야 할 때 유리할 수 있지만 신중한 처리가 필요합니다.

**Comment by le2sky** [comment]
https://velog.io/@kkambbak1/%EB%82%B4%EB%B6%80-%EB%B0%98%EB%B3%B5%EA%B3%BC-%EC%99%B8%EB%B6%80-%EB%B0%98%EB%B3%B5
https://hyune-c.tistory.com/45

**Comment by GIVEN53** [comment]
for문
외부 반복
명령형
작은 데이터셋 성능 높음, 명시적인 제어

stream
내부 반복
선언형
벙렬 스트림, 가독성
불변성을 위해 추가적인 메모리 사용
